### DISCUSSION TITLE
Internet Explorer 2.0.0 port help (just a spare time project)

### DISCUSSION THREAD

**Discussion by acalcutt** [maplibre/maplibre-gl-js#691] [General]
Hi Everyone,

In my project I still have some internet explorer compatibility, mainly because I'm a glutton for pain I think (haha). I was working on this more as a 'is it possible?' type project.

The other day I was messing with [maplibre-gl-inspect](https://github.com/acalcutt/maplibre-gl-inspect) and decided to see if I could make it internet explorer compatible, since my version of the plugin was not. I ended up finding with babel conversion as part of my build script, it worked pretty easily and I was able to add babelify to the build script an it just worked.

Because of how well the plugin change went, I got wondering how hard it would be to make a maplibre-gl-js 2.0.0 branch compatible with internet explorer again. I set off with very little knowledge of the build process, just following what the build script was doing and ended up in rollup ( rollup.config.js ).

In my initial attempt I tried to add babel transform using rollup plugins that replaced object.assign, fetch, and polyfill , but after several attempts I always ended up with a mosty working map, with my water layer missing and a style that wasn't completely loaded

![image](https://user-images.githubusercontent.com/3792408/144612410-846ae7f1-9ae1-4ea5-a196-8e65aa4eaf9c.png)

In my second attempt I started to understand the build process a little better. I had found 'build/rollup_plugins.js' and was starting to understand the build process. I knew 1.15.2 still works with internet explorer, so I wanted to look how it was done in that version. After looking I found it was mostly done by 'buble', which converts the javascript to ES5 level. I tried to put this back and was getting some errors with types. Initially I fixed the 'types' error by putting '[surcase](https://www.npmjs.com/package/sucrase)'  in front of buble (which worked), but after more testing I ended putting back the mapbox flow() function which seemed to remove types also. 

Adding buble back fixed the object.assign issues, like my babel test, but I still had to do something to fix fetch, polyfill, and EndWith , abort-controller errors that showed in Internet explorer. I fixed these by adding some code to src/index.ts to add some IE backward compatibility 

I added 
```
import 'isomorphic-fetch';
import "abort-controller/polyfill"
```
and
```
if (typeof String.prototype.endsWith !== 'function') {
    String.prototype.endsWith = function(suffix) {
        return this.indexOf(suffix, this.length - suffix.length) !== -1;
    };
}
```
This fixed most of the errors in Internet explorer, but unfortunately I ended up right where my babel attempt put me. The map mostly loads, but gets stuck at the water layer. Internet explorer does not give me any more errors (except two SCRIPT0 errors that seem to relate to a perfomance.now test that shouldn't matter).

The version I made works with no errors in other browsers, but in IE i just can't get past that water layer. You can see here ( https://wifidb.net/wifidb/opt/map.php?func=wifidbmap&labeled=0&test=1 **note the test=1 at the end**) my last test version is mostly working in IE, but works in other bowsers.

Unfortunately ,with no more useful errors being given I've run out of ideas on why this doesn't work. My latest code is up here ( https://github.com/acalcutt/maplibre-gl-js/tree/internet-explorer ) if anyone has any tips or ideas. 

This was just a spare time project, so I'm not to worried about getting this to work either way....but It feels close and I wonder why it doesn't work.

**Comment by HarelM** [comment]
I removed it due to security reason, lack of support and other issues.
I don't think it should be supported nowadays with win 10 and edge.
For users who would like to use maplibre-gl with IE version 1.15 is the right approach.
If you would like to keep support for IE and you would like to have new features in it I suggest to maintain the 1.x branch and merge to it whatever you like :-)
Going forward this is not the direction we would like to go...

**Comment by acalcutt** [comment]
i understand that. like i said, its a personal project for the fun of it in my own repository. not looking for it to be merged in, more to understand why it seems so close but is not working. feel free to ignore me ;)

**Comment by HarelM** [comment]
You can take a look at the changes made in the typescript PR in the rollup.config stuff.
Although after the removal of IE there were a few more commits to remove stuff that were needed only for IE.
bottom line, good luck!! :-)

**Reply by acalcutt** [reply]
Thanks, that's pretty much what I did for my second attempt at this. was hoping maybe I missed something obvious (probably something in those commit that remove the only IE stuff...haha)

It [mostly works](https://wifidb.net/wifidb/opt/map.php?func=wifidbmap&labeled=0&test=1). The layer I notice missing is my water layer, which is a fill/paint....so I was looking in code relating to that to see if there were any changes (or trying to), bit without an error it was hard find where to look.

**Comment by acalcutt** [comment]
I've been comparing the 1.x branch to main trying to find the things that have changed and I think I've reverted most of them here ( https://github.com/maplibre/maplibre-gl-js/compare/main...acalcutt:internetexplorer_test3 ). However I still end up at the same point as my other tests...the map mostly loads but won't display my water fill layer. If anyone has any ideas what has changed that I may have missed, I would definitely appreciate a push in the right direction. 

The typescript migration definitely makes it hard to compare 1.x to 2.x . there are just so many file name changes and flow comment removals it is just a lot of files. Through renaming the js to ts in the 1.x branch I was able to compare a lot, but it was still a LOT of changes. The structure looks like it has changed quite a bit with the removal of window.js, but it didn't seem like it should make a difference...but maybe it does idk.

**Comment by acalcutt** [comment]
I was finally able to get this working with Internet Explorer. This is 2.0.0 pre6 with the IE breaking changes reverted and few tweaks to make it ES5 javascript level. 
( https://github.com/acalcutt/maplibre-gl-js/tree/internetexplorer_test3 )

With that as a base is was easy to merge in other changes. For example, I made this version that includes the terrain3d branch
https://github.com/acalcutt/maplibre-gl-js/tree/2.x_es5_terrain

![image](https://user-images.githubusercontent.com/3792408/144887018-8529a11c-5c64-4b1a-a1a3-46a5c03e6b4e.png)
